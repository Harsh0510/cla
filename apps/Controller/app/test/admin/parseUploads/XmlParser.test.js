const Stream = require("stream");
const Readable = Stream.Readable;
const sax = require("sax");

const XmlParserRaw = require("../../../core/admin/parseUploads/XmlParser");

let openTagTimesCalled = 0;
let closeTagTimesCalled = 0;
let callbackHistory = [];

function getReadable(str) {
	const s = new Readable();
	s.push(str);
	s.push(null);
	return s;
}

async function parseXmlExplicit(saxOverride, str) {
	return await new XmlParserRaw(saxOverride).parse(getReadable(str));
}

async function parseXml(str) {
	return await parseXmlExplicit(sax, str);
}

beforeEach(function () {
	openTagTimesCalled = 0;
	closeTagTimesCalled = 0;
	callbackHistory = [];
});

test("invalid", async () => {
	let rootNode;
	let lastError;
	try {
		rootNode = await parseXml("<invalid></in>");
	} catch (e) {
		lastError = e;
	}
	expect(lastError.message).toMatch(/unexpected close tag/i);
	expect(rootNode).toBeUndefined();
});

test("empty xml", async () => {
	let rootNode = null;
	let lastError;
	try {
		rootNode = await parseXml("");
	} catch (e) {
		lastError = e;
	}
	expect(rootNode).toBeNull();
	expect(lastError).not.toBeNull();
});

test("xml tree", async () => {
	let rootNode;
	let lastError;
	try {
		rootNode = await parseXml(
			`
<?xml version="1.0" encoding="UTF-8"?>
<test>hello there<nested attri="bute"> nested <another/>tag here! </nested></test>
`.trim()
		);
	} catch (e) {
		lastError = e;
	}
	expect(lastError).toBeUndefined();
	expect(rootNode.type).toBe("tag");
	expect(rootNode.name).toBe("test");
	expect(rootNode.children.length).toBe(2);
	expect(rootNode.children[0].type).toBe("text");
	expect(rootNode.children[1].type).toBe("tag");
	expect(rootNode.children[1].name).toBe("nested");
	expect(rootNode.children[1].attribs).toEqual({ attri: "bute" });
	expect(rootNode).toMatchSnapshot();
});
