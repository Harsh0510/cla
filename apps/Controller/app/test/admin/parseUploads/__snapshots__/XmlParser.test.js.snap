// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xml tree 1`] = `
XmlNode {
  "attribs": Object {},
  "children": Array [
    XmlNode {
      "data": "hello there",
      "next": XmlNode {
        "attribs": Object {
          "attri": "bute",
        },
        "children": Array [
          XmlNode {
            "data": " nested ",
            "next": XmlNode {
              "attribs": Object {},
              "children": Array [],
              "name": "another",
              "next": XmlNode {
                "data": "tag here! ",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "text",
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
            },
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
          XmlNode {
            "attribs": Object {},
            "children": Array [],
            "name": "another",
            "next": XmlNode {
              "data": "tag here! ",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "text",
            },
            "parent": [Circular],
            "prev": XmlNode {
              "data": " nested ",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
            "type": "tag",
          },
          XmlNode {
            "data": "tag here! ",
            "next": null,
            "parent": [Circular],
            "prev": XmlNode {
              "attribs": Object {},
              "children": Array [],
              "name": "another",
              "next": [Circular],
              "parent": [Circular],
              "prev": XmlNode {
                "data": " nested ",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
              "type": "tag",
            },
            "type": "text",
          },
        ],
        "name": "nested",
        "parent": [Circular],
        "prev": [Circular],
        "type": "tag",
      },
      "parent": [Circular],
      "prev": null,
      "type": "text",
    },
    XmlNode {
      "attribs": Object {
        "attri": "bute",
      },
      "children": Array [
        XmlNode {
          "data": " nested ",
          "next": XmlNode {
            "attribs": Object {},
            "children": Array [],
            "name": "another",
            "next": XmlNode {
              "data": "tag here! ",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "text",
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
          },
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        XmlNode {
          "attribs": Object {},
          "children": Array [],
          "name": "another",
          "next": XmlNode {
            "data": "tag here! ",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "text",
          },
          "parent": [Circular],
          "prev": XmlNode {
            "data": " nested ",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
          "type": "tag",
        },
        XmlNode {
          "data": "tag here! ",
          "next": null,
          "parent": [Circular],
          "prev": XmlNode {
            "attribs": Object {},
            "children": Array [],
            "name": "another",
            "next": [Circular],
            "parent": [Circular],
            "prev": XmlNode {
              "data": " nested ",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
            "type": "tag",
          },
          "type": "text",
        },
      ],
      "name": "nested",
      "parent": [Circular],
      "prev": XmlNode {
        "data": "hello there",
        "next": [Circular],
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      "type": "tag",
    },
  ],
  "name": "test",
  "type": "tag",
}
`;
